jumanji/environments/packing/job_shop/improvement/get_actions.py

```diff
@@ -0,0 +1,296 @@
+
+    def body_fun(
+        loop_state: Tuple[jnp.int32, Tuple[chex.Array, chex.Array]],
+    ) -> Tuple[jnp.int32, Tuple[chex.Array, chex.Array]]:
+        """Update the critical block information based on the critical block pair.
+           During one update, if the pair (start_idx, end_idx) is valid, we update the
+           critical block information as follows:
+           - num_critical_block of end_idx is set to the block_id of the start_idx operation.
+           - left_end of end_idx is set to left_end of start_idx.
+           - is_right of start_idx is set to 0.
+           - right_neighbor of start_idx is set to the end_idx operation.
+           - is_left of end_idx is set to 0.
+           - left_neighbor of end_idx is set to the start_idx operation.
+
+        Args:
+            loop_state: Tuple containing (i, (critical_block_info, right_end_array))
+                   with i being the current index in the critical_block_pairs array
+
+        Returns:
+            Tuple containing (updated critical_block_info, updated right_end_array)
+        """
+        i, (critical_block_info, right_end_array) = loop_state
+        start_idx, end_idx = critical_block_pairs[i]
+
+        block_id = critical_block_info[start_idx, 1]
+        left_end_of_start_idx = critical_block_info[start_idx, 6]
+
+        critical_block_info = (
+            critical_block_info.at[end_idx, CBFields.BLOCK_ID]
+            .set(block_id)
+            .at[start_idx, CBFields.IS_RIGHT]
+            .set(0)
+            .at[start_idx, CBFields.RIGHT_NEIGHBOR]
+            .set(end_idx)
+            .at[end_idx, CBFields.LEFT_NEIGHBOR]
+            .set(start_idx)
+            .at[end_idx, CBFields.LEFT_END]
+            .set(left_end_of_start_idx)
+        )
+
+        right_end_array = right_end_array.at[block_id].set(end_idx)
+
+        return (i + 1, (critical_block_info, right_end_array))
+
+    init_state = (0, (critical_block_info, right_end_array))
+
+    _, (final_critical_block_info, right_end_array) = jax.lax.while_loop(
+        cond_fun, body_fun, init_state
+    )
+
+    # Set right_end of each operation to the right_end of its critical block
+    block_ids = final_critical_block_info[:, 1]
+    right_ends = jnp.where(block_ids != -1, right_end_array[block_ids], -1)
+    final_critical_block_info = final_critical_block_info.at[:, CBFields.RIGHT_END].set(right_ends)
+
+    # For each critical block, find left operations.
+    # Due to the initialization of num_critical_block, the left operation are the ones
+    # with index equal to num_critical_block.
+    final_critical_block_info = final_critical_block_info.at[:, CBFields.IS_LEFT].set(
+        final_critical_block_info[:, CBFields.BLOCK_ID] == jnp.arange(num_ops_total)
+    )
+
+    return final_critical_block_info
+
+
+def get_action_mask_n5(critical_block_info: chex.Array, max_num_ops: int) -> chex.Array:
+    """Get the mask of valid actions for the N5 neighborhood.
```

Comment:
Do we explain somewhere what N5/N6 means ? might be good to explain it somewhat or put a reference to some external document somewhere